### PR TITLE
Add WMT dataset

### PR DESCRIPTION
**Pull request by jaketae**
This PR supersedes #42 by creating a `WMTTask` class that inherits from from `AutoTask`. All functionality in the previous PR have been carried over. For convenience, the PR description has been copied and pasted from the original thread.

----
This PR implements the following:
* Load the English validation portion of WMT 19
* Load an AR model to run CLM, with `stride` tokens given as context
* Report perplexity, normalized by the number of tokens

Estimated runtime on GPU is ~1 minute; CPU, ~10 minutes.

### CODE REVIEW

**Comment by jaketae** [comment]
While I've checked that `python3 -m evaluation.eval  --model_name_or_path=gpt2  --eval_tasks wmt --output_dir outputs` works as intended, `tokenizer` complains that it has been forked by multiprocessing, presumably coming from the `Dataloader` and its `num_workers` argument. 

```shell
(venv) jaketae:evaluation (wmt2) $ python3 -m evaluation.eval  --model_name_or_path=gpt2  --eval_tasks wmt --output_dir outputs
08/19/2021 21:09:49 - evaluation - INFO - Beginning evaluation on device cpu
08/19/2021 21:09:49 - evaluation - INFO - Loading model...
08/19/2021 21:10:01 - evaluation - INFO - Benchmarking wmt...
Reusing dataset wmt19 (/Users/jaketae/.cache/huggingface/datasets/wmt19/kk-en/1.0.0/fae232cf0c13b62b26731bafce0810bc652fb5799189790bed836db0cee28056)
  0%|                                                                                                     | 0/14 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
[W ParallelNative.cpp:212] Warning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (function set_num_threads)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
[W ParallelNative.cpp:212] Warning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (function set_num_threads)
``

The script continues with this warning, but it is worthy of note.

File: evaluation/tasks/wmt/wmt.py [hunk 1]
```diff
@@ -1,2 +1,55 @@
...
+        )
+        loader = DataLoader(
+            dataset,
+            batch_size=self.task_config["batch_size"],
+            num_workers=self.task_config["num_workers"],
+            shuffle=False,
+            drop_last=True,
+        )
+        log_likelihoods = []
+        for input_ids in tqdm(loader):
```
Comment (tttyuntian) [inline evaluation/tasks/wmt/wmt.py:46]: A quick fix!

```suggestion
        for input_ids in tqdm(loader, desc=f"Evaluating {self.get_display_name()}"):
```

File: evaluation/tasks/wmt/wmt.py [hunk 2]
```diff
@@ -1,2 +1,55 @@
...
+            input_ids = input_ids.to(self.device)
+            target_ids = input_ids.clone()
+            target_ids[:, :-stride] = -100
+            with torch.no_grad():
+                outputs = self.model(input_ids, labels=target_ids)
+                log_likelihood = outputs[0]
+            log_likelihoods.append(log_likelihood)
+        perplexity = torch.exp(torch.stack(log_likelihoods).sum() / len(loader))
+        self.metrics["perplexity"] = perplexity.item()
```
Comment (tttyuntian) [inline evaluation/tasks/wmt/wmt.py:55]: Should we think about standardizing the `self.metrics` definition format? I personally prefer to do the below. Anyway, this is a minor change, so it's your call!

```suggestion
        self.metrics = {
                "perplexity": torch.exp(torch.stack(log_likelihoods).sum() / len(loader)).item(),
        }
```

File: evaluation/tasks/wmt/wmt.py [hunk 3]
```diff
@@ -1,2 +1,55 @@
...
+            batch_size=self.task_config["batch_size"],
+            num_workers=self.task_config["num_workers"],
+            shuffle=False,
+            drop_last=True,
+        )
+        log_likelihoods = []
+        for input_ids in tqdm(loader, desc=f"Evaluating {self.get_display_name()}"):
+            input_ids = input_ids.to(self.device)
+            target_ids = input_ids.clone()
+            target_ids[:, :-stride] = -100
```
Comment (wilsonyhlee) [inline evaluation/tasks/wmt/wmt.py:49]: nit: The code looks good. But for the sake of readability (in a public repo), I typically like to add some simple comments explaining what and why is being done in operations like these.

**Review by wilsonyhlee — COMMENTED** [review]
Looks good overall. Thanks for working on this dataset. Just have one very small nit comment.

**Review by tttyuntian — COMMENTED** [review]
Thanks for the work! The code is concise and clean. I dropped few comments on some code format related points.

File: evaluation/tasks/wmt/wmt.py (hunk 2 above)
Comment (jaketae) [reply evaluation/tasks/wmt/wmt.py:55]: Thank you for the suggestion! I also agree that it would be helpful to have a standardized format. But I think I'll stick to the version I have for now, because I think putting everything in the `self.metrics` re-initialization makes the code somewhat long. I hope this is okay! Thanks for the feedback nonetheless.

File: evaluation/tasks/wmt/wmt.py (hunk 3 above)
Comment (jaketae) [reply evaluation/tasks/wmt/wmt.py]: Thanks for the feedback! Just added a comment line above to indicate that it's for ignoring context tokens during loss computation. Hope this is fine!

**Comment by jaketae** [comment]
> The script continues with this warning, but it is worthy of note. Also, as somewhat expected, this warning does not appear if I do not pass any `num_worker` arguments into the `Dataloader`.

I am going with this solution for now, but I did leave a TODO note for posterity. I can see this becoming a problem, since the `tokenizer` + `Dataloader` combo will probably be used in many parts of the repo.

**Review by wilsonyhlee — APPROVED** [review]
LGTM! Thanks for the work!

**Comment by tianjianjiang** [comment]
> While I've checked that the script works as intended, the tokenizer complains that it has been forked by multiprocessing, which is presumably coming from the `Dataloader` and its `num_workers` argument.
> 
> ```
> (venv) jaketae:evaluation (wmt2) $ python3 -m evaluation.eval  --model_name_or_path=gpt2  --eval_tasks wmt --output_dir outputs
> 08/19/2021 21:09:49 - evaluation - INFO - Beginning evaluation on device cpu
> 08/19/2021 21:09:49 - evaluation - INFO - Loading model...
> 08/19/2021 21:10:01 - evaluation - INFO - Benchmarking wmt...
> Reusing dataset wmt19 (/Users/jaketae/.cache/huggingface/datasets/wmt19/kk-en/1.0.0/fae232cf0c13b62b26731bafce0810bc652fb5799189790bed836db0cee28056)
>   0%|                                                                                                     | 0/14 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
> To disable this warning, you can either:
>         - Avoid using `tokenizers` before the fork if possible
>         - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
> [W ParallelNative.cpp:212] Warning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (function set_num_threads)
> huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
> To disable this warning, you can either:
>         - Avoid using `tokenizers` before the fork if possible
>         - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
> [W ParallelNative.cpp:212] Warning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (function set_num_threads)
> ```
> 
> The script continues with this warning, but it is worthy of note. Also, as somewhat expected, this warning does not appear if I do not pass any `num_worker` arguments into the `Dataloader`.

Just a side note that you guys may already be aware of it. Besides the dedicated issues from HuggingFace repos, there's an interesting comment on an issue from AllenNLP: https://github.com/allenai/allennlp/issues/5128#issuecomment-830136469

(I totally forgot to comment on this until I ran into it again today.)

**Comment by jaketae** [comment]
Hey @tianjianjiang, thanks for the heads up. So what I'm getting out of that thread are:

1. The warning is fine and expected.
2. To suppress the warning, an environmental variable has to be set.

AFAIK, there is no safe and guaranteed way of setting an environment variable within a Python script, which means the only way to suppress the warning is for the user to set this variable in the terminal process before they run the evaluation. Is this the correct interpretation, or are there other alternatives?

**Comment by tianjianjiang** [comment]
Hey @jaketae 

Apologies for the late reply.

> 1. The warning is fine and expected.

As long as there's no deadlock then I think so.
Or a minor consideration: if just doing hyperparameter tuning, I prefer it to be fully deterministic for a fair comparison, such that no multiprocessing/threading will happen in my experiments.

> 2. To suppress the warning, an environmental variable has to be set.

I guess it may also be possible that we reexamine when we call the tokenizer.

> AFAIK, there is no safe and guaranteed way of setting an environment variable within a Python script, which means the only way to suppress the warning is for the user to set this variable in the terminal process before they run the evaluation. Is this the correct interpretation, or are there other alternatives?

I think so. Even if we try to use `os.environ`, in theory it still isn't bullet-proof.

I suppose the only true alternative is ensuring a perferrable call order if at all possible. Sometimes it isn't possible and yet it might be fine.
